genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/JobExecutionStateMachineImpl.java

```diff
@@ -18,102 +18,266 @@
-import org.springframework.statemachine.StateMachine;
-import org.springframework.statemachine.listener.StateMachineListenerAdapter;
+import org.springframework.boot.task.TaskExecutorBuilder;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
 
+import javax.annotation.Nullable;
+import java.util.Collection;
 import java.util.List;
-import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
- * Implementation of JobExecutionStateMachine.
+ * Implementation of the job execution state machine.
  *
  * @author mprimi
  * @since 4.0.0
  */
 @Slf4j
 public class JobExecutionStateMachineImpl implements JobExecutionStateMachine {
-
-    @Getter
-    private final StateMachine<States, Events> stateMachine;
-    private final StateMachineListenerAdapter<States, Events> executionCompletionListener;
-    private final CountDownLatch completionLatch = new CountDownLatch(1);
-    private final ExecutionContext executionContext;
+    private static final int RETRY_DELAY = 1000;
+    private final ThreadPoolTaskExecutor taskExecutor;
     @Getter
     private final List<ExecutionStage> executionStages;
+    private final ExecutionContext executionContext;
+    private final AtomicReference<Future<?>> runFuture = new AtomicReference<>();
+    private final JobExecutionListener listener;
+    private AtomicReference<States> currentState = new AtomicReference<>(States.READY);
 
     /**
      * Constructor.
      *
-     * @param stateMachine     the state machine
-     * @param executionContext the execution context
-     * @param executionStages  the execution stages
+     * @param executionStages  the (ordered) list of execution stages
+     * @param executionContext the execution context passed across stages during execution
+     * @param listeners        the list of listeners
      */
     public JobExecutionStateMachineImpl(
-        final StateMachine<States, Events> stateMachine,
+        final List<ExecutionStage> executionStages,
         final ExecutionContext executionContext,
-        final List<ExecutionStage> executionStages
+        final Collection<JobExecutionListener> listeners
     ) {
-        this.stateMachine = stateMachine;
         this.executionStages = executionStages;
         this.executionContext = executionContext;
-        this.stateMachine
-            .getExtendedState()
-            .getVariables()
-            .put(ExecutionStage.EXECUTION_CONTEXT_CONTEXT_KEY, executionContext);
-        this.executionCompletionListener = new StateMachineListenerAdapter<States, Events>() {
-            @Override
-            public void stateMachineStopped(final StateMachine<States, Events> sm) {
-                log.debug("State machine execution complete");
-                completionLatch.countDown();
-            }
-        };
+
+        this.taskExecutor = new TaskExecutorBuilder()
```

Comment:
I really wanted to be explicit on the fact that this runs on its own (single) thread.
Maybe rather than injecting it i could replace `start()` and `stop()` with a `run()` call which makes the executor superfluous.